### PR TITLE
RavenDB-14990 improve JSON serialization performance

### PR DESCRIPTION
**Pull request by ppekrol**

### CODE REVIEW

**Review by ayende — APPROVED** [review]

File: src/Raven.Client/Json/Converters/JsonDictionaryDateTimeKeysConverter.cs [hunk 1]
```diff
@@ -121,7 +123,12 @@ public override object ReadJson(JsonReader reader, Type objectType, object exist
 
         public override bool CanConvert(Type objectType)
         {
-            if (objectType.GetTypeInfo().IsGenericType == false)
+            return _cache.GetOrAdd(objectType, CanConvertInternal);
```
Comment (ayende) [inline src/Raven.Client/Json/Converters/JsonDictionaryDateTimeKeysConverter.cs:126]: This is expensive?

File: src/Raven.Client/Json/Converters/JsonEnumerableConverter.cs [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+                return false;
+
+            var jsonArrayContract = _conventions.JsonContractResolver.ResolveContract(objectType) as JsonArrayContract;
+            if (jsonArrayContract == null)
+                return false;
+
+            if (jsonArrayContract.IsMultidimensionalArray)
+                return false;
+
+            return jsonArrayContract.CollectionItemType != typeof(object);
```
Comment (ayende) [inline src/Raven.Client/Json/Converters/JsonEnumerableConverter.cs:47]: What is the type is an interface? `List<IFoo>` ?

**Review by ayende — APPROVED** [review]

File: src/Raven.Client/Json/Converters/JsonDictionaryDateTimeKeysConverter.cs (hunk 1 above)
Comment (ppekrol) [reply src/Raven.Client/Json/Converters/JsonDictionaryDateTimeKeysConverter.cs:126]: Original implementation took 1600ms in my test. This one takes 400ms.

File: src/Raven.Client/Json/Converters/JsonEnumerableConverter.cs (hunk 2 above)
Comment (ppekrol) [reply src/Raven.Client/Json/Converters/JsonEnumerableConverter.cs:47]: Need to double check, but shouldn't be an issue.

Comment (ppekrol) [reply src/Raven.Client/Json/Converters/JsonEnumerableConverter.cs:47]: Tested. It yields exactly the same JSON as before. Left code in the issue that I've used.

**Pull request merged**